From 1223df6504653b5297053015869a43f180e0e088 Mon Sep 17 00:00:00 2001
From: Erik Krogen <ekrogen@linkedin.com>
Date: Sun, 13 Nov 2016 14:50:33 -0800
Subject: [PATCH 2245/2848] HDFS-11087. NamenodeFsck should check if the
 output writer is still writable. Contributed by
 Erik Krogen.

(cherry picked from commit 98d9e9c4c823d865ddfc017e71c82d12c8328e53)
(cherry picked from commit c997fc629d9b89ba998decc56d410cf6341a71d9)

Change-Id: I970f0d623490b7d0c99e74fc51c7d0ceca8e2553
---
 .../hadoop/hdfs/server/namenode/NamenodeFsck.java  |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeFsck.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeFsck.java
index 277b1b3..3ab351e 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeFsck.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NamenodeFsck.java
@@ -500,7 +500,14 @@ void check(String parent, HdfsFileStatus file, Result res) throws IOException {
     } else {
       out.print('.');
     }
-    if (res.totalFiles % 100 == 0) { out.println(); out.flush(); }
+    if (res.totalFiles % 100 == 0) {
+      out.println();
+      // checkError here will attempt to flush the stream, or report an error
+      // if the stream has encountered an error or been closed by the client
+      if (out.checkError()) {
+        throw new IOException("fsck encountered an error in its output stream");
+      }
+    }
     int missing = 0;
     int corrupt = 0;
     long missize = 0;
-- 
1.7.9.5

